### PR TITLE
Add `ShadowBevel` component and mixin

### PR DESCRIPTION
**Pull request by aaronccasanova**
Part of https://github.com/Shopify/polaris-summer-editions/issues/711

This PR introduces a `ShadowBevel` component and `shadow-bevel` mixin.

The `shadow-bevel` mixin is intended for internal use when access to component markup is limited or would require significant refactoring. In many cases, component stylesheets are already applying `border`, `border-radius`, and `box-shadow` where the upgrade path is simply removing the decls and including the mixin (no need to modify any markup).

The `ShadowBevel` component is intended for internal ~and external~ usage. This is particularly useful with components that have been converted to use Layout components in which there is no stylesheet or class to attach the `shadow-bevel` mixin. Note: Consumers should get this behavior for free by nature of using Polaris components with the upcoming integration (e.g. in `Card`, `Popover`, etc.). ~However, the primitive is being exposed for custom implementations.~

**UPDATE:** This PR will [keep the `ShadowBevel` component internal](https://github.com/Shopify/polaris/commit/8de9956dce10be7be18824defb710922de64c780) until needed externally.

[Storybook](https://5d559397bae39100201eedc1-ecayaukxxk.chromatic.com/?path=/story/all-components-shadowbevel--default)

Example Sass upgrade:

```diff
.Banner {
-  border: var(--p-border-width-1) solid var(--p-color-border);
-  border-radius: var(--p-border-radius-3);
-  box-shadow: var(--p-shadow-md);
  
+ @include shadow-bevel(
+   $boxShadow: var(--p-shadow-md);
+   $borderRadius: var(--p-border-radius-3);  
+ );
}
```

Example TSX upgrade:

```diff
function Card(props) {
-  return <Box>...</Box>
+ return (
+   <ShadowBevel boxShadow="md" borderRadius="3">
+     <Box>...</Box>
+   </ShadowBevel>
+ )
}
```

Example usage:

https://github.com/Shopify/polaris/assets/32409546/cfb83b26-4c3a-4096-9a8a-6e5cabafcd98

### CODE REVIEW

File: polaris-react/src/index.ts [hunk 1]
```diff
@@ -329,6 +329,9 @@ export type {SelectAllActionsProps} from './components/SelectAllActions';
 export {SettingToggle} from './components/SettingToggle';
 export type {SettingToggleProps} from './components/SettingToggle';
 
+export {ShadowBevel} from './components/ShadowBevel';
```
Comment (sam-b-rose) [inline polaris-react/src/index.ts]: I do agree this should be publicly available for consumer to use this style.

Do we need to export this now or can we verify this internally first until we are ready to make it officially supported? Just might be nice to have a buffer of time to work out any potential bugs/gotchas

File: polaris-react/src/styles/foundation/_shadow-bevel.scss [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/// Applies a shadow-bevel effect to an element.
+/// @param $boxShadow - Polaris ShadowBorderAlias custom property.
+/// @param $borderRadius - Polaris BorderRadiusScale custom property.
+/// @param $content - Declaration passed to the pseudo-element. Used for toggling the bevel effect.
+/// @param $zIndex - Declaration passed to the pseudo-element. Used for adjusting the stacking order.
+@mixin shadow-bevel(
```
Comment (sam-b-rose) [inline polaris-react/src/styles/foundation/_shadow-bevel.scss:6]: Do we want this as a sharable mixin or could we only use the `ShadowBevel` component as our preferred method of style composition?

I'm thinking we could still have the mixin but keep it local to only the `ShadowBevel.scss` so other components are more likely to use the `ShadowBevel` component instead. Wdyt? @aaronccasanova

**Review by sam-b-rose — APPROVED** [review]
Nice work!! The bevels are looking awesome, can't wait to see them applied to the components 👏 Just a few thoughts and questions, but nothing blocking

File: polaris-react/src/index.ts (hunk 1 above)
Comment (aaronccasanova) [reply polaris-react/src/index.ts]: Agreed. Let's remove the export for now 👍

File: polaris-react/src/styles/foundation/_shadow-bevel.scss (hunk 2 above)
Comment (aaronccasanova) [reply polaris-react/src/styles/foundation/_shadow-bevel.scss:6]: > Do we want this as a sharable mixin or could we only use the `ShadowBevel` component as our preferred method of style composition?

I think having only one solution could introduce friction to the upgrade path. I mentioned in the PR description that the mixin is intended for use when access to component markup is limited or would require significant refactoring. For example, have a look through the hardcoded [`--p-font-*` custom properties](https://github.com/search?q=repo%3AShopify%2Fpolaris+path%3Apolaris-react+--p-font-&type=code) that can't easily upgrade to the `Text` component, but would benefit from this same strategy (e.g. `@include text-heading-sm`, `@include text-body-lg`, etc). Additionally, I would personally opt for the mixin before the component to [help reduce excessive DOM size](https://developer.chrome.com/en/docs/lighthouse/performance/dom-size/) costs.

> I'm thinking we could still have the mixin but keep it local to only the `ShadowBevel.scss` so other components are more likely to use the ShadowBevel component instead. Wdyt? @aaronccasanova

I also think this is appropriately positioned in the library. The effect is foundational to the system and it is customary to import these types of primitives from `styles/common`

File: polaris-react/src/index.ts (hunk 1 above)
Comment (aaronccasanova) [reply polaris-react/src/index.ts]: Done https://github.com/Shopify/polaris/pull/9638/commits/8de9956dce10be7be18824defb710922de64c780

File: polaris-react/src/styles/foundation/_shadow-bevel.scss (hunk 2 above)
Comment (sam-b-rose) [reply polaris-react/src/styles/foundation/_shadow-bevel.scss:6]: Sounds good to me! Thanks for the added context, examples, and rational 👍

**Review by sarahill — APPROVED** [review]
🎉  Nice work! I tried some different background colors and it's looking good

**Review by kyledurand — COMMENTED** [review]
In your description you mentioned you wanted the component to be public but it's not included in `src/index.ts`. Otherwise I think this looks good but I can take another look once it's getting bundled

File: polaris-react/src/components/ShadowBevel/ShadowBevel.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+import {ShadowBevel} from './ShadowBevel';
+
+export default {
+  component: ShadowBevel,
+} as ComponentMeta<typeof ShadowBevel>;
+
+export function Default() {
+  return (
+    <VerticalStack gap="5">
```
Comment (aveline) [inline polaris-react/src/components/ShadowBevel/ShadowBevel.stories.tsx:37]: Perhaps consider adding some stories with colored backgrounds to show the bevel on something other than a plain white background

**Review by aveline — APPROVED** [review]

**Comment by aaronccasanova** [comment]
> In your description you mentioned you wanted the component to be public but it's not included in src/index.ts

@kyledurand Sam and I [decided to keep the component internal](https://github.com/Shopify/polaris/pull/9638#discussion_r1254801785) until we have external use cases. Updated the PR description 👆

**Review by kyledurand — APPROVED** [review]

File: polaris-react/src/components/ShadowBevel/ShadowBevel.stories.tsx (hunk 3 above)
Comment (aaronccasanova) [reply polaris-react/src/components/ShadowBevel/ShadowBevel.stories.tsx:37]: Good call! Done https://github.com/Shopify/polaris/pull/9638/commits/289528f65d71c244b15f2becc18f8a5c2578d4ec

image

**Review by sophschneider — APPROVED** [review]
✨

**Pull request merged**